torch/csrc/jit/passes/quantization.cpp

```diff
@@ -1754,22 +1754,83 @@ std::tuple<at::Tensor, at::Tensor> FoldConvBatchNorm2dHelper::
   return std::make_tuple(new_w, new_b);
 }
 
+c10::optional<IValue> extractIValueFromGraph(
+    const script::Module& m,
+    const std::string& aten_op_name,
+    const size_t input_index) {
+  auto graph = m.get_method("forward").graph();
+
+  std::stack<Block*> blocks_to_visit;
+  blocks_to_visit.push(graph->block());
+  while (!blocks_to_visit.empty()) {
+    Block* b = blocks_to_visit.top();
+    blocks_to_visit.pop();
+    for (Node* n : b->nodes()) {
+      if (n->kind() == Symbol::aten(aten_op_name)) {
+        return toIValue(n->input(input_index));
+      }
+    }
+  }
+  return {};
+}
+
 bool FoldConvBatchNorm2dHelper::tryExtractingConvBNParameters(
     Module& conv,
     Module& bn,
     ConvBNParameters& r) {
   if (!hastensor(conv, "weight") || !conv.hasattr("bias") ||
-      !hastensor(bn, "weight") || !hastensor(bn, "bias") ||
-      !hastensor(bn, "running_mean") || !hastensor(bn, "running_var") ||
-      !bn.hasattr("eps")) {
+      !hastensor(bn, "running_mean") || !hastensor(bn, "running_var")) {
     return false;
   }
 
+  static constexpr size_t bn_weight_index = 1;
+  static constexpr size_t bn_bias_index = 2;
+  static constexpr size_t bn_running_mean_index = 3;
+  static constexpr size_t bn_running_var_index  = 4;
+  static constexpr size_t bn_eps_index = 7;
+  static constexpr size_t conv_bias_index = 2;
+
+  std::string bn_op_name("batch_norm");
+
   r.bn_rm = bn.attr("running_mean").toTensor();
   r.bn_rv = bn.attr("running_var").toTensor();
-  r.bn_eps = bn.attr("eps").toDouble();
-  r.bn_w = bn.attr("weight").toTensor();
-  r.bn_b = bn.attr("bias").toTensor();
+  if (bn.hasattr("eps")) {
+    r.bn_eps = bn.attr("eps").toDouble();
+  } else {
+    auto optional_eps = extractIValueFromGraph(bn, bn_op_name, bn_eps_index);
+    if (!optional_eps) {
+      return false;
+    }
+    r.bn_eps = optional_eps.value().toDouble();
+  }
+  r.bn_w = at::ones_like(r.bn_rm);
+  if (bn.hasattr("weight")) {
+    if (!bn.attr("weight").isNone()) {
+      r.bn_w = bn.attr("weight").toTensor();
+    }
+  } else {
+    auto optional_bn_weight = extractIValueFromGraph(bn, bn_op_name, bn_weight_index);
+    if (!optional_bn_weight) {
+      return false;
+    }
+    if (!optional_bn_weight.value().isNone()) {
+      r.bn_w = optional_bn_weight.value().toTensor();
+    }
+  }
+  r.bn_b = at::zeros_like(r.bn_rm);
+  if (bn.hasattr("bias")) {
+    if (!bn.attr("bias").isNone()) {
+      r.bn_b = bn.attr("bias").toTensor();
+    }
+  } else {
+    auto optional_bn_bias = extractIValueFromGraph(bn, bn_op_name, bn_bias_index);
+    if (!optional_bn_bias) {
+      return false;
+    }
+    if (!optional_bn_bias.value().isNone()) {
```

Comment:
actually, can you use pattern match to extract these values?

Reply:
I think this is what we can do, you'll have a pattern match that matches the `batch_norm` call, and then try to look at each input values and try to extract the IValue. when there is no attribute, then you can use these extracted values

Reply:
Sure. Note that the value extraction is only done when the corresponding attributes are not present in the module.

Reply:
we can always extract values, and always look at attribute first. something like following:
```
extraced_bn_params = extractConstantBnValues(...)
if (bn.hasattr("bias")) {
  ...
} else {
 // look at extraced_bn_params
}
```

Reply:
Sounds good.